File: Day5/main.py
from time import perf_counter_ns
from copy import copy


def part1(rules, checks):
    result = 0
    for line in checks:
        nums = [int(x) for x in line.split(",")]
        valid = True
        for before, after in rules:
            if before in nums or after in nums:
                try:
                    before_index = nums.index(before)
                except Exception:
                    before_index = None
                try:
                    after_index = nums.index(after)
                except Exception:
                    after_index = None
                if before_index is not None and after_index is not None:
                    if before_index > after_index:
                        valid = False
        if valid:
            result += nums[int(len(nums) / 2)]
    return result


def part2(rules, checks):
    result = 0
    for line in checks:
        nums = [int(x) for x in line.split(",")]
        corrected = False
        while True:
            for before, after in rules:
                if before in nums or after in nums:
                    try:
                        before_index = nums.index(before)
                    except Exception:
                        before_index = None
                    try:
                        after_index = nums.index(after)
                    except Exception:
                        after_index = None
                    if before_index is not None and after_index is not None:
                        if before_index > after_index:
                            before_num = copy(nums[before_index])
                            after_num = copy(nums[after_index])
                            nums[before_index] = after_num
                            nums[after_index] = before_num
                            corrected = True
                            break
            else:
                break
        if corrected:
            result += nums[int(len(nums) / 2)]
    return result


def main():
    data = open("example.txt", "r").read().splitlines()
    rules = []
    checks = []
    for line in data:
        if "|" in line:
            x, y = line.split("|")
            rules.append((int(x), int(y)))
        elif "," in line:
            checks.append(line)
    print(rules)
    print(checks)
    start_time = perf_counter_ns()
    part_1_solution = part1(rules, checks)
    print(f"Part 1: {part_1_solution}")
    part_2_solution = part2(rules, checks)
    print(f"TTC part 1: {perf_counter_ns() - start_time}ns")
    print(f"Part 2: {part_2_solution}")
    print(f"TTC: {perf_counter_ns() - start_time}ns")


if __name__ == "__main__":
    main()

File: Day5/src/main.rs
use std::env;
use std::fs::File;
use std::io::prelude::*;
use std::path::Path;
use std::time::Instant;

fn read_file(file_name: &String) -> String {
    // Create a path to the desired file
    let path = Path::new(&file_name);
    let display = path.display();

    // Open the path in read-only mode
    let mut file = match File::open(&path) {
        Err(why) => panic!("could not open {}: {}", display, why),
        Ok(file) => file,
    };

    // Read the file contents into a string
    let mut s = String::new();
    match file.read_to_string(&mut s) {
        Err(why) => panic!("could no open file {}: {}", display, why),
        Ok(_) => file,
    };
    return s;
}

fn part1(rules: Vec<(i32, i32)>, checks: Vec<&str>) -> i32 {
    let mut result = 0;
    for check in checks {
        let nums: Vec<&str> = check.split(",").collect();
        let values: Vec<i32> = nums.iter().map(|&v| v.parse::<i32>().unwrap()).collect();
        let mut valid: bool = true;
        for (before, after) in &rules {
            if values.contains(before) || values.contains(after) {
                let mut before_index = 0;
                match &values.iter().position(|b| b == before) {
                    Some(index) => before_index = index.clone(),
                    None => continue,
                };
                let mut after_index: usize = 0;
                match &values.iter().position(|a| a == after) {
                    Some(index) => after_index = index.clone(),
                    None => continue,
                };
                if before_index > after_index {
                    valid = false;
                }
            }
        }
        if valid == true {
            result += values[values.len() / 2]
        }
    }
    result
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        panic!("File name is required")
    }
    let file_name = &args[1];
    let data = read_file(file_name);
    let mut rules: Vec<(i32, i32)> = Vec::new();
    let mut checks: Vec<&str> = Vec::new();
    for line in data.lines() {
        if line.contains("|") {
            let values: Vec<&str> = line.split("|").collect();
            let value1 = match values[0].parse::<i32>() {
                Err(_) => panic!("invalid number"),
                Ok(num) => num,
            };
            let value2 = match values[1].parse::<i32>() {
                Err(_) => panic!("invalid number"),
                Ok(num) => num,
            };
            rules.push((value1, value2))
        } else if line.contains(",") {
            checks.push(line);
        }
    }
    let part_1 = part1(rules, checks);
    let start = Instant::now();
    assert_eq!(part_1, 143);
    println!("Part 1 {}", part_1);
    let duration = start.elapsed();
    println!("TTC: {:?}ns", duration.as_nanos());
}
